tests/test_passes/test_stub.py

```diff
@@ -0,0 +1,63 @@
+import pytest
+
+import magma as m
+from magma.passes.stub import StubPass
+from magma.testing import check_files_equal
+
+
+# NOTE(rsetaluri): We need this fixture per function, since each function may
+# run a pass to modify the circuits. However, the setup is identical for each so
+# we reuse this fixture across all the test.
+@pytest.fixture(scope="function")
+def ckts():
+
+    class _Foo(m.Circuit):
+        io = m.IO(I=m.In(m.Bit), O=m.Out(m.Bit))
+        io.O @= io.I
+
+
+    class _Bar(m.Circuit):
+        io = m.IO(I=m.In(m.Bit), O=m.Out(m.Bit))
+        io.O @= io.I
+
+
+    class _Top(m.Circuit):
+        io = m.IO(I=m.In(m.Bit), O=m.Out(m.Bit))
+        io.O @= _Foo()(io.I)
+
+    return (_Foo, _Bar, _Top)
+
+
+def test_noop(ckts):
+    _Foo, _Bar, _Top = ckts
+    m.compile("build/test_passes_stub_noop", _Top, output="coreir")
+    assert check_files_equal(__file__,
+                             f"build/test_passes_stub_noop.json",
+                             f"gold/test_passes_stub_noop.json")
+
+
+def test_basic(ckts):
+    _Foo, _Bar, _Top = ckts
+    StubPass(_Top, (_Foo,)).run()
+    m.compile("build/test_passes_stub_basic", _Top, output="coreir")
+    assert check_files_equal(__file__,
+                             f"build/test_passes_stub_basic.json",
+                             f"gold/test_passes_stub_basic.json")
+
+
+def test_unused(ckts):
+    _Foo, _Bar, _Top = ckts
+    StubPass(_Top, (_Bar,)).run()
+    m.compile("build/test_passes_stub_unused", _Top, output="coreir")
+    assert check_files_equal(__file__,
+                             f"build/test_passes_stub_unused.json",
+                             f"gold/test_passes_stub_noop.json")
+
+
+def test_compiler_args(ckts):
+    _Foo, _Bar, _Top = ckts
+    m.compile("build/test_passes_stub_compiler_args", _Top,
+              output="coreir", stubs=(_Foo,))
+    assert check_files_equal(__file__,
+                             f"build/test_passes_stub_compiler_args.json",
+                             f"gold/test_passes_stub_basic.json")
```

Comment:
Can you add a test that stubs a circuit with an internal instance?  I believe that you also need to remove those (since the compilation algorithm will iterate over the instances and generate the coreir for them)

Reply:
This is a good point. In fact, this breaks the abstraction that we shouldn't be deleting from circuits, and it looks like all downstream compilers place even floating instances. One option would be to make the compilers all deal with floating instances differently (not place them), then all we would have to do is unwire everything. However, this changes the semantics of the compiler and may affect other codepaths.

Instead, I think stubbing/black-boxing should be ephemeral -- the circuits should not be modified and only the invocation of the compiler for which these options are passed should consider those options. The only downside is that each compiler (verilog, coreir, etc.) would have to duplicate the logic to deal with stubs/black-boxes (rather than have this fall out of the modified circuits). I think this is still better though, esp. since we really only use CoreIR. I'll modify the PR to do this.